<a href="https://colab.research.google.com/github/a00351269/Actividades_Aprendizaje-/blob/main/Actividad6DS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ** Actividad Semanal -- 6, visualización**
## **Maestría en Inteligencia Artificial Aplicada**
### **Curso: TC4029.10 Ciencia y analítica de datos (Gpo 10)**
## **Tecnológico de Monterrey**
## **Prof: Dra, Maria de la Paz Rico Fernandez**



### **Alumno: Genaro Ramos Higuera  -  Matrícula A00351269**
### **Fecha: 31 de octubre del 2022**

1.- Descarga los datos y carga el dataset en tu libreta. 

2.- Obten la información del DataFrame con los métodos y propiedades: shape, columns, head(), dtypes, info(), isna()

3.- Limpia los datos eliminando los registros nulos o rellena con la media de la columna

4.- Calcula la estadística descriptiva con describe() y explica las medidas de tendencia central y dispersión

5.- Realiza el conteo de las variables categóricas

6.- Escala los datos, si consideras necesario

7.- Reduce las dimensiones con PCA, si consideras necesario.

7.1.- Indica la varianza de los datos explicada por cada componente seleccionado. Para actividades de exploración de los datos la varianza > 70%

7.2.-Indica la importancia de las variables en cada componente

8.- Elabora los histogramas de los atributos para visualizar su distribución

9.- Realiza la visualización de los datos usando por lo menos 3 gráficos que consideres adecuados: plot, scatter, jointplot, boxplot, areaplot, pie chart, pairplot, bar chart, etc.

10.- Interpreta y explica cada uno de los gráficos indicando cuál es la información más relevante que podría ayudar en el proceso de toma de decisiones.







#1.-Descarga los datos y carga el dataset en tu libreta.

Comenzamos por importar las librerías para trabajar

In [66]:
import pandas as pd
import numpy as np

Proseguimos convirtiendo el archivo csv a trabajar a una data frame

In [79]:
df = pd.read_csv('https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv')
df.head()

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15      X16  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0      0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   3455.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0  14948.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0  28959.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0  19146.0   

       X17     X18      X19      X20     X21     X22     X23    Y  
0      0.0     0.0    689.0      0.0     0.0     0.0     0.0  1.0  
1   3261.0     0.0   1000.0   1000.0  1000.0     0.0  2000.0  1.0  
2  15549.0  1518.0   1500.0   1000.0  1000.0  1000.0  5000.0  0.0  
3  29547.0  2000.0   2019.0   1200.0  1100.0  1069.0  1000.0  0.0  
4  19131.0  2000.0  36681.0  10000.0  9000.0   689.0   679.0  0.0  

[5 rows x 25 columns]

#2.- Obten la información del DataFrame con los métodos y propiedades: shape, columns, head(), dtypes, info(), isna()

Esta es la informacion del Dataframe que tiene 30,000 registros, y notamos que algunos tienen registros con algunas columnas con datos nulos:

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

In [4]:
df.shape

(30000, 25)

In [5]:
df.dtypes

ID       int64
X1       int64
X2     float64
X3     float64
X4     float64
X5     float64
X6     float64
X7     float64
X8     float64
X9     float64
X10    float64
X11    float64
X12    float64
X13    float64
X14    float64
X15    float64
X16    float64
X17    float64
X18    float64
X19    float64
X20    float64
X21    float64
X22    float64
X23    float64
Y      float64
dtype: object

Creamos un diccionario con significado cada columna del dataframe.
Y con ello poder decidir que hacer con los datos faltantes.

In [ ]:
columns_name = ['BAL','SEX','EDUCATION','MARRIAGE','AGE','PAYSTAT_1','PAYSTAT_2','PAYSTAT_3','PAYSTAT_4','PAYSTAT_5','PAYSTAT_6','BILL_1','BILL_2','BILL_3','BILL_4','BILL_5','BILL_6','PAY_1',
                'PAY_2','PAY_3','PAY_4','PAY_5','PAY_6','Y']

In [6]:
Meanings_col= ['ID',
              'Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit. ',
              'Gender (1 = male; 2 = female)',
              'Education (1 = graduate school; 2 = university; 3 = high school; 4 = others)',
              'Marital status (1 = married; 2 = single; 3 = others)',
              'Age (year)',
              'the repayment status in September, 2005, The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above',
              'the repayment status in August, 2005',
              'the repayment status in July, 2005',
              'the repayment status in June, 2005',
              'the repayment status in May, 2005',
              'the repayment status in April, 2005',
              'amount of bill statement in September, 2005',
              'amount of bill statement in August, 2005',
              'amount of bill statement in July, 2005',
              'amount of bill statement in June, 2005',
              'amount of bill statement in May, 2005',
              'amount of bill statement in April, 2005',
              'amount paid in September, 2005',
              'amount paid in August, 2005',
              'amount paid in July, 2005',
              'amount paid in June, 2005',
              'amount paid in May, 2005',
              'amount paid in April, 2005',
              'the real probability of default as the response variable']
meanings = {}
n = 0
for col in df.columns:
  meanings[col] = Meanings_col[n]
  n = n + 1
meanings


{'ID': 'ID',
 'X1': 'Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit. ',
 'X2': 'Gender (1 = male; 2 = female)',
 'X3': 'Education (1 = graduate school; 2 = university; 3 = high school; 4 = others)',
 'X4': 'Marital status (1 = married; 2 = single; 3 = others)',
 'X5': 'Age (year)',
 'X6': 'the repayment status in September, 2005, The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above',
 'X7': 'the repayment status in August, 2005',
 'X8': 'the repayment status in July, 2005',
 'X9': 'the repayment status in June, 2005',
 'X10': 'the repayment status in May, 2005',
 'X11': 'the repayment status in April, 2005',
 'X12': 'amount of bill statement in September, 2005',
 'X13': 'amount of bill statement in August, 2005',
 'X14': 'amount of bill sta

#3.- Limpia los datos eliminando los registros nulos o rellena con la media de la columna


Primeramente verificamos que en nuestro archivo no se encuentren ID repetidos, y esto lo hacemos con las siguientes instrucciones:

In [7]:
if df['ID'].count() == df["ID"].unique().size:
  print ("Tenemos los siguientes registros ID únicos: ", df["ID"].unique().size)
else: 
  print ("Existen registros algunos repetidos y solo tenemos registros ID únicos: ", df["ID"].unique().size)

Tenemos los siguientes registros ID únicos:  30000


In [8]:
#Saquemos un reporte de los registros nulos.
missing_data = df.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

ID
False    30000
Name: ID, dtype: int64

X1
False    30000
Name: X1, dtype: int64

X2
False    29999
True         1
Name: X2, dtype: int64

X3
False    29998
True         2
Name: X3, dtype: int64

X4
False    29998
True         2
Name: X4, dtype: int64

X5
False    29995
True         5
Name: X5, dtype: int64

X6
False    29997
True         3
Name: X6, dtype: int64

X7
False    29995
True         5
Name: X7, dtype: int64

X8
False    29993
True         7
Name: X8, dtype: int64

X9
False    29991
True         9
Name: X9, dtype: int64

X10
False    29984
True        16
Name: X10, dtype: int64

X11
False    29986
True        14
Name: X11, dtype: int64

X12
False    29989
True        11
Name: X12, dtype: int64

X13
False    29989
True        11
Name: X13, dtype: int64

X14
False    29987
True        13
Name: X14, dtype: int64

X15
False    29985
True        15
Name: X15, dtype: int64

X16
False    29983
True        17
Name: X16, dtype: int64

X17
False    29990
True        10
Name: X17, dt

Ahora veamos los valores de Monto de Crédito Otorgados que sean válidos.

Verificamos  que X1 incluyan solo montos positivos :

In [9]:
df.loc[df['X1'] < 0] # para saber que registros tienen un monto negativo

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

Seguido, checamos que X2 solo sean valores de 1 o 2

In [11]:
df_clean=df.copy()

In [12]:
(df_clean['X2'].values < 1).sum() + (df_clean['X2'].values > 2).sum()

0

In [13]:
hombres=df_clean.loc[df_clean['X2'] == 1,"X2"].count()
mujeres=df_clean.loc[df_clean['X2'] == 2,"X2"].count()
prop_hombres=hombres/(hombres+mujeres)
prop_mujeres=mujeres/(hombres+mujeres)

print ("De los cuales hombres son ",hombres," y mujeres ", mujeres)
print ("Con proporcion de hombres de ", prop_hombres, "y de mujeres de ", prop_mujeres, " respectivamente.")

De los cuales hombres son  11887  y mujeres  18112
Con proporcion de hombres de  0.39624654155138506 y de mujeres de  0.6037534584486149  respectivamente.


Sabemos que hay 30,000 registros esto significa que uno de ellos es Nulo

In [16]:
# creamos una serie booleana para encontrar el registro nulo
sexo_nulo = pd.isnull(df_clean["X2"])
   
# filtramos los datos
# desplegamos el datos con "X2" = NaN
df_clean[sexo_nulo]


ID      X1  X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  X18  \
24365  24366  130000 NaN NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN  NaN   

       X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

In [17]:
#Ahora reeemplazamos el registro con sexo nulo dandole el 2 por ser el mayor porcentaje con el 60%
df_clean["X2"].fillna(2,inplace=True)
df_clean.loc[24365:24365]

ID      X1   X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  \
24365  24366  130000  2.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN   

       X18  X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

Ahora seguimos con X3, que son el nivel de educación solo debe admitir valores de 1, 2, 3, y 4 para cualquier otro. Así que verificamos

In [18]:
print('Valores menores a 1: ' + str((df_clean['X3'].values < 1).sum()))
print('Valores mayores a 4: ' + str((df_clean['X3'].values > 4).sum()))

Valores menores a 1: 14
Valores mayores a 4: 331


Vemos que hay 331 valores mayores a 4. En este caso tomaremos la suposición de que todo valor mayor a 4 puede entrar dentro del valor 4, que significa "4 = others".

In [19]:
X3_values = []
for i in range(0,len(df_clean.index)):
  if df_clean['X3'].iloc[i] > 4:
    if df_clean['X3'].iloc[i] not in X3_values:
      X3_values.append(int(df_clean['X3'].iloc[i]))
for i in X3_values:
  df_clean['X3'].replace(i,4,inplace=True)

print('Valores mayores a 4: ' + str((df_clean['X3'].values > 4).sum()))

Valores mayores a 4: 0


Mientras que los valores menores a 1, que solo son 14, los remplazaremos el promedio

In [20]:
X3_mean = df_clean.X3.mean()
print('Promedio de X3: ' + str(int(X3_mean)))
df_clean['X3'].replace(0,int(X3_mean),inplace=True)
print('Valores menores a 1: ' + str((df_clean['X3'].values < 1).sum()))

Promedio de X3: 1
Valores menores a 1: 0


In [21]:
# creamos una serie booleana para encontrar registros nulos en X3
educacion_nulos = pd.isnull(df_clean["X3"])
   
# filtramos los datos
# desplegamos los registros con "X3" = NaN
df_clean[educacion_nulos]

ID      X1   X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16      X17  \
6227    6228   30000  1.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  21120.0   
24365  24366  130000  2.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN      NaN   

          X18     X19     X20     X21     X22    X23    Y  
6227   1586.0  1365.0  1663.0  5106.0  1050.0    0.0  0.0  
24365     NaN     NaN     NaN     NaN     NaN  390.0  0.0  

[2 rows x 25 columns]

In [22]:
# les asignamos la media de educacion = 1 a los 2 registros nulos encontrados
df_clean["X3"].fillna(1,inplace=True)

In [23]:
df_clean.loc[24365:24365]


ID      X1   X2   X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  \
24365  24366  130000  2.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN   

       X18  X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

In [24]:
df_clean.loc[6227:6227]

ID     X1   X2   X3  X4  X5  X6  X7  X8  X9  ...  X15  X16      X17  \
6227  6228  30000  1.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  21120.0   

         X18     X19     X20     X21     X22  X23    Y  
6227  1586.0  1365.0  1663.0  5106.0  1050.0  0.0  0.0  

[1 rows x 25 columns]

No debiesen haber registros nulos en Educacion X3 y lo comprobamos así:

In [26]:
educacion_nulos = pd.isnull(df_clean["X3"])
   
# filtramos los datos
# desplegamos los registros con "X3" = NaN
df_clean[educacion_nulos] 

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

X4 solo debería admitir valores de 1, 2, y 3. así que verificamos

In [27]:
print('Valores menores a 1: ' + str((df_clean['X4'].values < 1).sum()))
print('Valores mayores a 3: ' + str((df_clean['X4'].values > 3).sum()))

Valores menores a 1: 54
Valores mayores a 3: 0


Descubrimos 54 valores de 0, así que sustituimos por el promedio

In [28]:
X4_mean = df_clean.X4.mean()
print('Promedio de X4: ' + str(int(X4_mean)))
df_clean['X4'].replace(0,int(X4_mean),inplace=True)
print('Valores menores a 1: ' + str((df_clean['X4'].values < 1).sum()))

Promedio de X4: 1
Valores menores a 1: 0


In [29]:
# creamos una serie booleana para encontrar registros nulos en X4
marital_status_nulos = pd.isnull(df_clean["X4"])
   
# filtramos los datos
# desplegamos los registros con "X4" = NaN
df_clean[marital_status_nulos]

ID      X1   X2   X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  \
6227    6228   30000  1.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN   
24365  24366  130000  2.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN   

           X17     X18     X19     X20     X21     X22    X23    Y  
6227   21120.0  1586.0  1365.0  1663.0  5106.0  1050.0    0.0  0.0  
24365      NaN     NaN     NaN     NaN     NaN     NaN  390.0  0.0  

[2 rows x 25 columns]

In [30]:
# les asignamos la media de Marital Status = 1 Married a los 2 registros nulos encontrados
df_clean["X4"].fillna(1,inplace=True)

In [31]:
df_clean.loc[6227:6227, "ID":"X4"]

ID     X1   X2   X3   X4
6227  6228  30000  1.0  1.0  1.0

In [32]:
df_clean.loc[24365:24365, "ID":"X4"]

ID      X1   X2   X3   X4
24365  24366  130000  2.0  1.0  1.0

Ahora checamos X5. Donde no debería haber menores a 18

In [33]:
print('Valores menores a 18: ' + str((df_clean['X5'].values < 18).sum()))

Valores menores a 18: 0


In [34]:
# creamos una serie booleana para encontrar registros nulos en X5
edad_nulos = pd.isnull(df_clean["X5"])
   
# filtramos los datos
# desplegamos los registros con "X5" = NaN
df_clean[edad_nulos]

ID      X1   X2   X3   X4  X5   X6   X7   X8   X9  ...       X15  \
182      183  500000  2.0  1.0  1.0 NaN  0.0  0.0  0.0  0.0  ...  122967.0   
219      220  310000  2.0  1.0  2.0 NaN -1.0 -1.0 -1.0 -1.0  ...       0.0   
239      240  140000  2.0  2.0  3.0 NaN  0.0  0.0  0.0  0.0  ...   19068.0   
6227    6228   30000  1.0  1.0  1.0 NaN  NaN  NaN  NaN  NaN  ...       NaN   
24365  24366  130000  2.0  1.0  1.0 NaN  NaN  NaN  NaN  NaN  ...       NaN   

            X16      X17      X18      X19      X20      X21      X22  \
182    108834.0  70064.0  70010.0  30357.0  30000.0  20000.0  52183.0   
219         0.0      0.0   4542.0    126.0      0.0      0.0      0.0   
239     16409.0  16383.0   3000.0   2000.0   2198.0   1000.0   3000.0   
6227        NaN  21120.0   1586.0   1365.0   1663.0   5106.0   1050.0   
24365       NaN      NaN      NaN      NaN      NaN      NaN      NaN   

           X23    Y  
182    20000.0  0.0  
219        0.0  0.0  
239     2399.0  1.0  
6227       0.0  0.0  
24365    390.0  0.0  

[5 rows x 25 columns]

In [35]:
promedio_edad=int(df_clean['X5'].mean())
promedio_edad

35

Reemplazamos los 5 registros nulos por la edad promedio de 35 años, ya que apenas son el 0.016% del total de registros.

In [36]:
df_clean["X5"].fillna(promedio_edad,inplace=True) 

In [37]:
df_clean.loc[182:182, "ID":"X5"]

ID      X1   X2   X3   X4    X5
182  183  500000  2.0  1.0  1.0  35.0

In [38]:
df_clean.loc[219:219, "ID":"X5"]

ID      X1   X2   X3   X4    X5
219  220  310000  2.0  1.0  2.0  35.0

In [39]:
df_clean.loc[239:239, "ID":"X5"]

ID      X1   X2   X3   X4    X5
239  240  140000  2.0  2.0  3.0  35.0

In [40]:
df_clean.loc[6227:6227, "ID":"X5"]

ID     X1   X2   X3   X4    X5
6227  6228  30000  1.0  1.0  1.0  35.0

In [41]:
df_clean.loc[24365:24365, "ID":"X5"]

ID      X1   X2   X3   X4    X5
24365  24366  130000  2.0  1.0  1.0  35.0

Para X6-X11, no deben haber valores menores a -1, ni mayores a 9

In [42]:
print('Valores menores a -1: ' + str((df_clean.iloc[:,6:12].values < -1).sum()))
print('Valores mayores a 9: ' + str((df_clean.iloc[:,6:12].values > 9).sum()))

Valores menores a -1: 24415
Valores mayores a 9: 0


In [43]:
print('Encontramos que son iguales a -2: ' + str((df_clean.iloc[:,6:12].values == -2).sum()))

Encontramos que son iguales a -2: 24415


En este caso, asumiremos que los Valores Menores a -1, esto es -2, equivalen a -1 y se asume o que son errores de captura, o significan que llevan periodos consecutivos cumpliendo. Ya que -1 dice que se pagó a tiempo, y cualquier valor mayor a ese son los meses que se han atrasado en pagos.

In [44]:
df_clean.iloc[:,6:12] = df_clean.iloc[:,6:12].replace(-2,-1)
print('Valores menores a -1: ' + str((df_clean.iloc[:,6:12].values < -1).sum()))

Valores menores a -1: 0


In [45]:
(df_clean.iloc[:,6:7].values == 0).sum()

14736

Ahora veamos a los registros con valores NULOS en las Columnas X6 a X11


In [46]:
keys=["X6","X7","X8","X9","X10","X11"]

for i in keys:
  nulos = pd.isnull(df_clean[i])
  print(df_clean[nulos])
  print("Mediana ", df_clean[i].median())
  print("Registros NO NULOS: ", df_clean[i].count())

          ID      X1   X2   X3   X4    X5  X6  X7  X8  X9  ...  X15  X16  \
6227    6228   30000  1.0  1.0  1.0  35.0 NaN NaN NaN NaN  ...  NaN  NaN   
24123  24124   50000  1.0  1.0  2.0  29.0 NaN NaN NaN NaN  ...  NaN  NaN   
24365  24366  130000  2.0  1.0  1.0  35.0 NaN NaN NaN NaN  ...  NaN  NaN   

           X17     X18     X19     X20     X21     X22    X23    Y  
6227   21120.0  1586.0  1365.0  1663.0  5106.0  1050.0    0.0  0.0  
24123      NaN     NaN     NaN     NaN     NaN     NaN  360.0  0.0  
24365      NaN     NaN     NaN     NaN     NaN     NaN  390.0  0.0  

[3 rows x 25 columns]
Mediana  0.0
Registros NO NULOS:  29997
          ID      X1   X2   X3   X4    X5   X6  X7  X8  X9  ...  X15  X16  \
6227    6228   30000  1.0  1.0  1.0  35.0  NaN NaN NaN NaN  ...  NaN  NaN   
24123  24124   50000  1.0  1.0  2.0  29.0  NaN NaN NaN NaN  ...  NaN  NaN   
24365  24366  130000  2.0  1.0  1.0  35.0  NaN NaN NaN NaN  ...  NaN  NaN   
29832  29833   20000  1.0  2.0  2.0  30.0  0.0 N

In [47]:
keys=["X6","X7","X8","X9","X10","X11"] #imputamos la mediana a las casillas nulas de cada columna X6 a la X11
copiadf=df_clean
for i in keys:
  nulos = pd.isnull(copiadf[i])
  mediana=(copiadf[i].median())
  copiadf[i].fillna(mediana,inplace=True)

In [48]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      30000 non-null  float64
 3   X3      30000 non-null  float64
 4   X4      30000 non-null  float64
 5   X5      30000 non-null  float64
 6   X6      30000 non-null  float64
 7   X7      30000 non-null  float64
 8   X8      30000 non-null  float64
 9   X9      30000 non-null  float64
 10  X10     30000 non-null  float64
 11  X11     30000 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

Para X12-X23 checamos que no existan valores negativos

In [49]:
print('Valores menores a 0: ' + str((df_clean.iloc[:,12:24].values < 0).sum()))

Valores menores a 0: 3932


En este caso, asumiremos que es un error de captura y son valores positivos. En la realidad tendríamos que verificar como se capturaron los datos para asegurarnos que sea posible hacer esto, o entender porque existen valores negativos y qué hacer con ellos

In [50]:
for col in df_clean.iloc[:,12:24].columns:
  df_clean[col] = df_clean[col].abs()
print('Valores menores a 0: ' + str((df_clean.iloc[:,12:24].values < 0).sum()))

Valores menores a 0: 0


In [52]:
keys=["X12","X13","X14","X15","X16","X17", "X18", "X19","X20","X21","X22", "X23"]

for i in keys:
  nulos = pd.isnull(df_clean[i])
  print(df_clean[nulos])
  print("Mediana ", df_clean[i].median())
  print("Registros NO NULOS: ", df_clean[i].count())

          ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15  X16  \
18        19  360000  2.0  1.0  1.0  49.0  1.0 -1.0 -1.0 -1.0  ...  NaN  NaN   
6227    6228   30000  1.0  1.0  1.0  35.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   
24123  24124   50000  1.0  1.0  2.0  29.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   
24365  24366  130000  2.0  1.0  1.0  35.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   
29733  29734  300000  1.0  2.0  1.0  37.0 -1.0 -1.0 -1.0 -1.0  ...  NaN  NaN   
29734  29735  500000  1.0  2.0  1.0  49.0 -1.0 -1.0  2.0 -1.0  ...  NaN  NaN   
29735  29736   50000  1.0  3.0  1.0  40.0  2.0  0.0  0.0  0.0  ...  NaN  NaN   
29824  29825   40000  1.0  1.0  1.0  47.0  2.0  2.0  2.0  2.0  ...  NaN  NaN   
29825  29826   50000  1.0  2.0  1.0  41.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   
29832  29833   20000  1.0  2.0  2.0  30.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   
29833  29834   20000  1.0  2.0  2.0  31.0 -1.0  0.0  0.0  0.0  ...  NaN  NaN   

           X17     X18     X19     X20 

In [55]:
#imputamos la mediana a las casillas nulas de cada columna X12 a la X23
keys=["X12","X13","X14","X15","X16","X17", "X18", "X19","X20","X21","X22", "X23"] 
copiadf=df_clean
for i in keys:
  nulos = pd.isnull(copiadf[i])
  mediana=(copiadf[i].median())
  copiadf[i].fillna(mediana,inplace=True)

In [56]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      30000 non-null  float64
 3   X3      30000 non-null  float64
 4   X4      30000 non-null  float64
 5   X5      30000 non-null  float64
 6   X6      30000 non-null  float64
 7   X7      30000 non-null  float64
 8   X8      30000 non-null  float64
 9   X9      30000 non-null  float64
 10  X10     30000 non-null  float64
 11  X11     30000 non-null  float64
 12  X12     30000 non-null  float64
 13  X13     30000 non-null  float64
 14  X14     30000 non-null  float64
 15  X15     30000 non-null  float64
 16  X16     30000 non-null  float64
 17  X17     30000 non-null  float64
 18  X18     30000 non-null  float64
 19  X19     30000 non-null  float64
 20  X20     30000 non-null  float64
 21  X21     30000 non-null  float64
 22

Vemos como quedó el dataframe limpio de nulos en las variables de entrada y hay 3 datos nulos en la variable de salida Y.

"Y" es la probabilidad del modelo. 
Dado que estos son datos faltantes no al azar , y como son pocos los datos nulos , 3 renglones , eliminaremos esas filas de los datos.


Procederemos a eliminar esos 3 renglones con datos nulos en la variable de salida:

In [61]:
df2 = df_clean.copy()
df2.dropna(inplace = True)

Verificamos que efectivamente no existan valores nulos

In [62]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29997 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      29997 non-null  int64  
 1   X1      29997 non-null  int64  
 2   X2      29997 non-null  float64
 3   X3      29997 non-null  float64
 4   X4      29997 non-null  float64
 5   X5      29997 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29997 non-null  float64
 8   X8      29997 non-null  float64
 9   X9      29997 non-null  float64
 10  X10     29997 non-null  float64
 11  X11     29997 non-null  float64
 12  X12     29997 non-null  float64
 13  X13     29997 non-null  float64
 14  X14     29997 non-null  float64
 15  X15     29997 non-null  float64
 16  X16     29997 non-null  float64
 17  X17     29997 non-null  float64
 18  X18     29997 non-null  float64
 19  X19     29997 non-null  float64
 20  X20     29997 non-null  float64
 21  X21     29997 non-null  float64
 22

Y como resultado tenemos una nueva data frame de 29997 filas y 25 columnas donde no existe ningún valor nulo

#4.- Funcion describe()


In [65]:
df2.describe()

ID              X1            X2            X3            X4  \
count  29997.000000    29997.000000  29997.000000  29997.000000  29997.000000   
mean   14999.803847   167496.072274      1.603794      1.840751      1.553689   
std     8659.837419   129748.803871      0.489116      0.743247      0.518326   
min        1.000000    10000.000000      1.000000      1.000000      1.000000   
25%     7501.000000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.000000   140000.000000      2.000000      2.000000      2.000000   
75%    22499.000000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      4.000000      3.000000   

                 X5            X6            X7            X8            X9  \
count  29997.000000  29997.000000  29997.000000  29997.000000  29997.000000   
mean      35.483782      0.075208     -0.007734     -0.030336     -0.075874   
std        9.217348      0.990661      1.035672      1.023859      0.987239   
min       21.000000     -1.000000     -1.000000     -1.000000     -1.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...            X15            X16            X17            X18  \
count  ...   29997.000000   29997.000000   29997.000000   29997.000000   
mean   ...   43339.381571   40385.523986   39017.091976    5662.352002   
std    ...   64285.015119   60750.283359   59462.986458   16562.848369   
min    ...       0.000000       0.000000       0.000000       0.000000   
25%    ...    2400.000000    1868.000000    1361.000000    1000.000000   
50%    ...   19068.000000   18128.000000   17135.000000    2100.000000   
75%    ...   54560.000000   50209.000000   49252.000000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

                X19            X20            X21            X22  \
count  2.999700e+04   29997.000000   29997.000000   29997.000000   
mean   5.921706e+03    5225.052405    4826.365170    4799.417042   
std    2.304194e+04   17607.010539   15666.756595   15278.899254   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    8.360000e+02     390.000000     298.000000     254.000000   
50%    2.009000e+03    1800.000000    1500.000000    1500.000000   
75%    5.000000e+03    4505.000000    4013.000000    4031.000000   
max    1.684259e+06  896040.000000  621000.000000  426529.000000   

                 X23             Y  
count   29997.000000  29997.000000  
mean     5216.012201      0.221189  
std     17778.281538      0.415054  
min         0.000000      0.000000  
25%       118.000000      0.000000  
50%      1500.000000      0.000000  
75%      4000.000000      0.000000  
max    528666.000000      1.000000  

[8 rows x 25 columns]